sotodlib/site_pipeline/update_smurf_caldbs.py

```diff
@@ -109,15 +123,252 @@ def smurf_detset_info(config:str | dict,
+    detsets = ctx.obsfiledb.get_detsets(obs_id)
+    min_ct = obs['start_time'] - 3600*24*7
+    cal_all = ctx.obsdb.query(
+        f"""
+        start_time <= {obs['start_time']} and subtype=='{cal_type}'
+        and start_time > {min_ct}
+        """, sort=['start_time']
+    )[::-1]
+
+    obs_ids = {
+        ds: None for ds in detsets
+    }
+
+    for o in cal_all:
+        dsets = ctx.obsfiledb.get_files(o['obs_id']).keys()
+        for ds in dsets:
+            if ds in obs_ids:
+                if obs_ids[ds] is None:
+                    obs_ids[ds] = o['obs_id']
+
+    return obs_ids
+
+# Dtype for calibration set
+cal_dtype = [
+    ('dets:readout_id', '<U40'),
+    ('dets:cal.r_tes', float),
+    ('dets:cal.r_frac', float),
+    ('dets:cal.p_bias', float),
+    ('dets:cal.s_i', float),
+    ('dets:cal.bg', int),
+    ('dets:cal.bg_polarity', int),
+    ('dets:cal.r_n', float),
+    ('dets:cal.p_sat', float),
+]
+
+@dataclass
+class CalInfo:
+    # Fields must be ordered like cal_dtype!
+    readout_id: str = ''
+
+    # From bias steps
+    r_tes: float = np.nan   # Ohm
+    r_frac: float = np.nan
+    p_bias: float = np.nan  # J
+    s_i: float = np.nan     # 1/V
+
+    # From IV
+    bg: int = -1
+    polarity: int = 1
+    r_n: float = np.nan  
+    p_sat: float = np.nan   # J
+
+
+def _load_smurf_npy(obs_id, substr):
+    """
+    Loads npy file from Z_smurf archive of book.
+
+    Args
+    _____
+    obs_id: str
+        obs-id of book to load file from
+    substr: str
+        substring to use to find numpy file in Z_smurf
+    """
+    files = ctx.obsfiledb.get_files(obs_id)
+    book_dir = os.path.dirname(list(files.values())[0][0][0])
+    smurf_dir = os.path.join(book_dir, 'Z_smurf')
+    for f in os.listdir(smurf_dir):
+        if substr in f:
+            fpath = os.path.join(smurf_dir, f)
+            break
+    else:
+        raise FileNotFoundError("Could not find npy file")
+    res = np.load(fpath, allow_pickle=True).item()
+    return res
+
+
+def get_cal_resset(ctx: core.Context, obs_id):
+    """Returns calibration ResultSet for a given ObsId"""
+    am = ctx.get_obs(obs_id, samples=(0, 1), ignore_missing=True)
+
+    cals = [CalInfo(rid) for rid in am.det_info.readout_id]
+
+    iv_obsids = get_cal_obsids(ctx, obs_id, 'iv')
+
+    rtm_bit_to_volt = None
+    ivas = {dset: None for dset in iv_obsids}
+    for dset, oid in iv_obsids.items():
+        if oid is not None:
+            ivas[dset] = _load_smurf_npy(oid, 'iv')
+            if rtm_bit_to_volt is None:
+                rtm_bit_to_volt = ivas[dset]['meta']['rtm_bit_to_volt']
+
+    bias_step_obsids = get_cal_obsids(ctx, obs_id, 'bias_steps')
+    bsas = {dset: None for dset in bias_step_obsids}
+    for dset, oid in bias_step_obsids.items():
+        if oid is not None:
+            bsas[dset] = _load_smurf_npy(oid, 'bias_step_analysis')
+            if rtm_bit_to_volt is None:
+                rtm_bit_to_volt = bsas[dset]['meta']['rtm_bit_to_volt']
```

Comment:
what's the point of getting `rtm_bit_to_volt` from the bsa and iva if you just overwrite it in line 249?

Reply:
oops was missing an if statement